engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala

```diff
@@ -85,79 +91,119 @@ class EngineJobExecutionActor(jobKey: BackendJobDescriptorKey,
-    case Event(hashes: CallCacheHashes, EmptyPartialCompletionData) =>
-      log.info(s"Got hashes for ${jobKey.mkTag(workflowId)}, awaiting job completion")
-      stay using PartialCompletionDataWithHashes(Success(hashes))
+    case Event(hashes: CallCacheHashes, data: SucceededResponseData) =>
+      saveCacheResults(hashes, data)
+    case Event(hashes: CallCacheHashes, data: NoResponseData) =>
+      addHashesAndStay(data, hashes)
 
-    case Event(HashError(t), PartialCompletionDataWithSucceededResponse(response)) =>
+    case Event(HashError(t), data: SucceededResponseData) =>
       recordHashError(t)
-      saveJobCompletionToJobStore(response)
-    case Event(HashError(t), EmptyPartialCompletionData) =>
+      saveJobCompletionToJobStore(data)
+    case Event(HashError(t), data: NoResponseData) =>
       recordHashError(t)
-      stay using PartialCompletionDataWithHashes(Failure(t))
-
-    case Event(response: SucceededResponse, PartialCompletionDataWithHashes(Success(hashes))) if effectiveCallCachingMode.writeToCache =>
-      saveCacheResults(CacheWriteOnCompletionData(response, hashes))
-    case Event(response: SucceededResponse, EmptyPartialCompletionData) if effectiveCallCachingMode.writeToCache =>
-      log.info(s"Got job result for {}, awaiting hashes", jobKey.mkTag(workflowId))
-      stay using PartialCompletionDataWithSucceededResponse(response)
-    case Event(response: BackendJobExecutionResponse, _) =>
-      saveJobCompletionToJobStore(response)
+      val updatedData = data.copy(hashes = Option(Failure(t)))
+      stay using updatedData
+
+    case Event(response: SucceededResponse, data @ NoResponseData(_, _, Some(Success(hashes)))) if effectiveCallCachingMode.writeToCache =>
+      saveCacheResults(hashes, response.updatedData(data))
+    case Event(response: SucceededResponse, data: NoResponseData) if effectiveCallCachingMode.writeToCache =>
+      log.info(s"Got job result for {}, awaiting hashes", jobTag)
+      stay using response.updatedData(data)
+    case Event(response: BackendJobExecutionResponse, data: NoResponseData) =>
+      saveJobCompletionToJobStore(response.updatedData(data))
   }
 
-  // When WritingToCallCache, the FSM always has EJEASuccessfulCompletionDataWithHashes
+  // When UpdatingCallCache, the FSM always has SucceededResponseData.
   when(UpdatingCallCache) {
-    case Event(CallCacheWriteSuccess, CacheWriteOnCompletionData(response, _)) =>
-      saveJobCompletionToJobStore(response)
-    case Event(CallCacheWriteFailure(reason), CacheWriteOnCompletionData(response, _)) =>
-      context.parent ! FailedNonRetryableResponse(jobKey, reason, response.returnCode)
+    case Event(CallCacheWriteSuccess, data: SucceededResponseData) =>
+      saveJobCompletionToJobStore(data)
+    case Event(CallCacheWriteFailure(reason), data: SucceededResponseData) =>
+      context.parent ! FailedNonRetryableResponse(jobKey, reason, data.successResponse.returnCode)
```

Comment:
@geoffjentry (and pinging @kcibul atop his mountain) is failing the job here really what we want?